src/Microsoft.ML.Core/Data/Repository.cs

```diff
@@ -113,14 +113,19 @@ internal Repository(bool needDir, IExceptionContext ectx)
             PathMap = new ConcurrentDictionary<string, string>();
             _open = new List<Entry>();
             if (needDir)
-                DirTemp = GetShortTempDir();
+                DirTemp = GetShortTempDir(ectx);
             else
                 GC.SuppressFinalize(this);
         }
 
-        private static string GetShortTempDir()
+        private static string GetShortTempDir(IExceptionContext ectx)
         {
-            var path = Path.Combine(Path.GetFullPath(Path.GetTempPath()), "ml_dotnet", Path.GetRandomFileName());
+            string path;
+            if (ectx is IHostEnvironmentInternal iHostInternal)
+                path = Path.Combine(Path.GetFullPath(iHostInternal.TempFilePath), "ml_dotnet", Path.GetRandomFileName());
+            else
+                path = Path.Combine(Path.GetFullPath(Path.GetTempPath()), "ml_dotnet", Path.GetRandomFileName());
+
             Directory.CreateDirectory(path);
             return path;
```

Comment:
This could be rewritten so it doesn't duplicate the strings:

```suggestion
            string tempPath = ectx is IHostEnvironmentInternal iHostInternal ?
                iHostInternal.TempFilePath :
                Path.GetTempPath();

            string path = Path.Combine(Path.GetFullPath(tempPath), "ml_dotnet", Path.GetRandomFileName());
            Directory.CreateDirectory(path);
            return path;
```